### ISSUE TITLE
General tips!

### ISSUE THREAD

**Issue by mickgortenmulder** [thelocomotion/GloriousProtonManager#1]
* Convert to a Python module to make it Object Oriented: https://packaging.python.org/en/latest/tutorials/packaging-projects/ (Or check out my sent aws-connector example)
* https://github.com/thelocomotion/GloriousProtonManager/blob/main/glorious_proton_manager.py#L11 Setting a home variable is something that should go into the init method, in the init. The init method is part of a Class and is called everytime you initiate the Class. Which works for this kind of variable setting, variables in the init method can be used everywhere in the Class. (Read upon class vs instance variables: https://medium.com/python-features/class-vs-instance-variables-8d452e9abcbd#:~:text=Class%20variables%20are%20shared%20across,surprising%20behaviour%20in%20our%20code  
* https://github.com/thelocomotion/GloriousProtonManager/blob/main/glorious_proton_manager.py#L16 You’re using the requests call twice. If you’re using something twice, put it in a separate method and call it by adding a variable into it. In this case create a method that does a requests.get, and insert the URL to get in the method as a variable.
* https://github.com/thelocomotion/GloriousProtonManager/blob/main/glorious_proton_manager.py#L23 I think unconsciously you’re doing a pretty good thing here if you consider this: When creating functions, you always need to limit what the function returns. This is a basic principle of object oriented coding. So by using the If and Else here you’re limiting what comes out of it. You can extend this function by adding the return statement. (read upon: https://realpython.com/python-return-statement/) . So try to use this as much as possible, use a try catch for special functions like http gets, and use if else for creating a ‘good’ and ‘bad’ outcome of your function. That way it would be also much easier to create unit tests, which brings to my next topic!
* I’m missing unit tests: https://docs.python.org/3/library/unittest.html (but you should reconsider if it’s really needed for this tool, it certainly pushed you to prevent writing broken or bad code)
* Separate the python functions from the command line code (PySimpleGUI), in a Python package the functions go into the lib or src folder and the command line code (PySimpleGUI) would then go into the bin folder.
* https://github.com/thelocomotion/GloriousProtonManager/blob/main/glorious_proton_manager.py#L19 this dict is something you would want to check in a method, so even see if there is a [1] (second array entry) and then pull the browser_download_url as this might not always be the case. You can also use scan functions to check for dict key values in a big dict and put a try and catch around or at least an if/else.

**Comment by PySimpleGUI** [comment]
![image](https://user-images.githubusercontent.com/46163555/177808304-10ac8f74-11a1-482a-bcb6-014f819ed1cb.png)

Would it be helpful to have any of these in your readme?

**Comment by thelocomotion** [comment]
Unfortunately I haven't had time to check all of the points brought up by @mickgortenmulder . I'll look into them ASAP and also add the most relevant points to the README

**Comment by thelocomotion** [comment]
@mickgortenmulder 

- All variables that matter were moved to constants.py. I read that they can also go to `__init__.py` but this does it for now. Unless there is a benefit to moving them to init that I don't know about other than having one less file to worry about.
- I reworked the functions to be as lean as possible, reducing the amount of code and especially if statements that were part of them. Now they basically do 'that one task' they are meant to do and nothing else. The if/else statements were moved to main() in `interface.py`. I also added comments to them. I'll look into following the advice you made though.
- Regarding the unit tests: Now that I got most things working again after the refactoring, I'll start doing some research on how it's done and put some in place
- The functions and main PySimpleGUI is now divided in separate files

Thanks for your comment!
